agent/command/s3_copy.go

```diff
@@ -75,105 +95,140 @@ type s3Loc struct {
-		}
+	}
 
-		err := validateS3BucketName(s3CopyFile.Source.Bucket)
-		if err != nil {
-			return errors.Wrapf(err, "source bucket '%v' is invalid",
-				s3CopyFile.Source.Bucket)
-		}
+	return catcher.Resolve()
+}
 
-		err = validateS3BucketName(s3CopyFile.Destination.Bucket)
-		if err != nil {
-			return errors.Wrapf(err, "destination bucket '%v' is invalid",
-				s3CopyFile.Destination.Bucket)
-		}
+// Apply the expansions from the relevant task config (including restricted expansion)
+// to all appropriate fields of the s3put.
+func (s3pc *s3copy) expandParams(conf *internal.TaskConfig) error {
+	var err error
+	if err = util.ExpandValues(s3pc, conf.GetExpansionsWithRestricted()); err != nil {
+		return errors.WithStack(err)
 	}
-
 	return nil
 }
 
-// Execute carries out the s3copy command - this is required
-// to satisfy the 'Command' interface
-func (c *s3copy) Execute(ctx context.Context,
+// Implementation of Execute.  Expands the parameters, and then copies the
+// resource from one s3 bucket to another one.
+func (s3pc *s3copy) Execute(ctx context.Context,
 	comm client.Communicator, logger client.LoggerProducer, conf *internal.TaskConfig) error {
 
-	// expand the S3 copy parameters before running the task
-	if err := util.ExpandValues(c, conf.Expansions); err != nil {
+	// expand necessary params
+	if err := s3pc.expandParams(conf); err != nil {
 		return errors.WithStack(err)
 	}
+	// re-validate command here, in case an expansion is not defined
+	if err := s3pc.validate(); err != nil {
+		return errors.WithStack(err)
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	s3pc.taskdata = client.TaskData{ID: conf.Task.Id, Secret: conf.Task.Secret}
 
 	errChan := make(chan error)
 	go func() {
-		errChan <- errors.WithStack(c.s3Copy(ctx, comm, logger, conf))
+		errChan <- errors.WithStack(s3pc.copyWithRetry(ctx, comm, logger, conf))
 	}()
 
 	select {
 	case err := <-errChan:
-		return errors.WithStack(err)
+		return err
 	case <-ctx.Done():
-		logger.Execution().Info("Received signal to terminate execution of S3 copy command")
+		logger.Execution().Info("Received signal to terminate execution of S3 Put Command")
 		return nil
 	}
+
 }
 
-// S3Copy is responsible for carrying out the core of the S3CopyPlugin's
-// function - it makes an API calls to copy a given staged file to it's final
-// production destination
-func (c *s3copy) s3Copy(ctx context.Context,
+func (s3pc *s3copy) copyWithRetry(ctx context.Context,
 	comm client.Communicator, logger client.LoggerProducer, conf *internal.TaskConfig) error {
 
-	td := client.TaskData{ID: conf.Task.Id, Secret: conf.Task.Secret}
+	timer := time.NewTimer(0)
+	defer timer.Stop()
+
+	td := s3pc.taskdata
 
 	var foundDottedBucketName bool
 
-	for _, s3CopyFile := range c.S3CopyFiles {
+	client := utility.GetHTTPClient()
+	client.Timeout = 10 * time.Minute
+	defer utility.PutHTTPClient(client)
+
+	ctx, cancel := context.WithCancel(context.Background())
```

Comment:
We do this in Execute and copyWithRetry; do we need it in both places?

Reply:
no :)